DiscordChatExporter.Core/Exporting/Filtering/Parsing/FilterParser.cs

```diff
@@ -0,0 +1,76 @@
+﻿using Superpower;
+using Superpower.Model;
+using Superpower.Parsers;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Text;
+
+namespace DiscordChatExporter.Core.Exporting.Filtering.Parsing
+{
+    public static class FilterParser
+    {
+        public static TokenListParser<FilterToken, string> AnyString { get; } =
+            Token.EqualTo(FilterToken.QuotedString).Apply(FilterTextParsers.QuotedString)
+            .Or(Token.EqualTo(FilterToken.UnquotedString).Apply(FilterTextParsers.UnquotedString));
+
+        public static TokenListParser<FilterToken, MessageFilter> AnyFilter { get; } =
+            from minus in Token.EqualTo(FilterToken.Minus).Optional()
+            from content in KeyValuePair.Or(TextFilter.Try()).Or(GroupedFilter)
+            select minus.HasValue ? new NegatedMessageFilter(content) : content;
+
+        public static TokenListParser<FilterToken, MessageFilter> KeyValuePair { get; } =
+            from key in AnyString.Try()
+            from colon in Token.EqualTo(FilterToken.Colon).Try()
+            from value in AnyString
+            select MessageFilter.CreateFilter(key, value);
+
+        public static TokenListParser<FilterToken, MessageFilter> TextFilter { get; } =
+            from value in AnyString
+            select MessageFilter.CreateFilter(value);
+
+        public static TokenListParser<FilterToken, MessageFilter> MultipleFilter { get; } =
+            from first in AnyFilter
+            from rest in Token.EqualTo(FilterToken.VBar)
+                .IgnoreThen(AnyFilter)
+                .Select(value => (Value: value, Unioned: true)) //anonymous struct keeps track of whether a vertical bar was used
+                .Or(AnyFilter.Select(value => (Value: value, Unioned: false)))
+                .Many()
+            select rest.Aggregate(first, (current, next) => new BooleanMessageFilter(current, next.Value, next.Unioned));
+
+        public static TokenListParser<FilterToken, MessageFilter> GroupedFilter { get; } =
+            from open in Token.EqualTo(FilterToken.LParen)
+            from content in MultipleFilter
+            from close in Token.EqualTo(FilterToken.RParen)
+            select content;
+
+        public static TokenListParser<FilterToken, MessageFilter> Instance { get; } = MultipleFilter.AtEnd();
+
+        public static bool TryParse(string input, [MaybeNullWhen(false)] out MessageFilter value, [MaybeNullWhen(true)] out string error, out Position errorPosition)
```

Comment:
Let's just do `public static MessageFilter Parse(string input)` and throw an exception directly in case of failure (instead of having to do it at the upper layers)

Reply:
For now I just entirely removed it and put the logic within the MessageFilter class since it uses both the tokenizer and parser. If we do remove the tokenizer, we can move it back if you'd like.